go-controller/hybrid-overlay/pkg/controller/master.go

```diff
@@ -0,0 +1,263 @@
+				klog.Errorf("failed to sync node %s hybrid overlay NodeSubnet annotation: %v", node.Name, err)
+			}
+		}
+		return nil
+	}
+
+	// Do not allocate a subnet if the node already has one
+	if extNodeSubnet != nil {
+		return nil
+	}
+
+	// No subnet reserved; allocate a new one
+	hostsubnetStr, err := m.allocator.AllocateNetwork()
+	if err != nil {
+		return fmt.Errorf("Error allocating hybrid overlay HostSubnet for node %s: %v", node.Name, err)
+	}
+	klog.Infof("Allocated hybrid overlay HostSubnet %s for node %s", hostsubnetStr, node.Name)
+
+	if _, _, err := net.ParseCIDR(hostsubnetStr); err != nil {
+		return fmt.Errorf("Error parsing hostsubnet %s: %v", hostsubnetStr, err)
+	}
+
+	if err := annotator.SetWithFailureHandler(types.HybridOverlayNodeSubnet, hostsubnetStr, func(node *kapi.Node, key string, val interface{}) {
+		if _, cidr, _ := net.ParseCIDR(val.(string)); cidr != nil {
+			_ = m.releaseNodeSubnet(node.Name, cidr)
+		}
+	}); err != nil {
+		return fmt.Errorf("failed to set node %s hybrid overlay NodeSubnet annotation: %v", node.Name, err)
+	}
+	return nil
+}
+
+func (m *MasterController) releaseNodeSubnet(nodeName string, subnet *net.IPNet) error {
+	err := m.allocator.ReleaseNetwork(subnet.String())
+	if err != nil {
+		return fmt.Errorf("Error deleting hybrid overlay HostSubnet %s for node %q: %s", subnet, nodeName, err)
+	}
+	klog.Infof("Deleted hybrid overlay HostSubnet %s for node %s", subnet, nodeName)
+	return nil
+}
+
+func (m *MasterController) handleOverlayPort(node *kapi.Node, annotator kube.Annotator) error {
+	// Only applicable to Linux nodes
+	if houtil.IsWindowsNode(node) {
+		return nil
+	}
+
+	_, haveDRMACAnnotation := node.Annotations[types.HybridOverlayDrMac]
+
+	subnet, err := ovn.ParseNodeHostSubnet(node)
+	if subnet == nil || err != nil {
+		// No subnet allocated yet; clean up
+		if haveDRMACAnnotation {
+			m.deleteOverlayPort(node)
+			annotator.Delete(types.HybridOverlayDrMac)
+		}
+		return nil
+	}
+
+	if haveDRMACAnnotation {
+		// already set up; do nothing
+		return nil
+	}
+
+	portName := houtil.GetHybridOverlayPortName(node.Name)
+	portMAC, portIP, _ := util.GetPortAddresses(portName)
+	if portMAC == nil || portIP == nil {
+		if portMAC == nil {
+			portMAC, _ = net.ParseMAC(util.GenerateMac())
+		}
+		if portIP == nil {
+			// Get the 3rd address in the node's subnet; the first is taken
+			// by the k8s-cluster-router port, the second by the management port
+			first := util.NextIP(subnet.IP)
+			second := util.NextIP(first)
+			portIP = util.NextIP(second)
+		}
+
+		klog.Infof("creating node %s hybrid overlay port", node.Name)
+
+		var stderr string
+		_, stderr, err = util.RunOVNNbctl("--", "--may-exist", "lsp-add", node.Name, portName,
+			"--", "lsp-set-addresses", portName, portMAC.String()+" "+portIP.String())
+		if err != nil {
+			return fmt.Errorf("failed to add hybrid overlay port for node %s"+
+				", stderr:%s: %v", node.Name, stderr, err)
```

Comment:
If someone tries to enable hybrid overlay on a cluster where one of the .2 IPs is already in use, it will just repeatedly keep trying to claim it and then logging errors. It would be nicer to just have it fail decisively.

Reply:
@danwinship so in our Slack discussion we decided that we'd just ignore this, and make our cluster network operator prevent changing hybrid overlay at runtime.